unleash-yggdrasil/src/lib.rs

```diff
@@ -94,84 +158,144 @@ impl EngineState {
+    fn harvest_metrics(&mut self) -> Option<MetricBucket> {
+        if let Some(state) = &self.compiled_state {
+            let metrics: HashMap<String, ToggleStats> = state
+                .values()
+                .map(|toggle| {
+                    let yes = toggle.yes.swap(0, Ordering::Relaxed);
+                    let no = toggle.no.swap(0, Ordering::Relaxed);
+
+                    let mut variants: HashMap<String, u32> = toggle
+                        .variants
+                        .iter()
+                        .map(|variant| {
+                            (
+                                variant.name.clone(),
+                                variant.count.swap(0, Ordering::Relaxed),
+                            )
+                        })
+                        .collect();
+
+                    variants.insert(
+                        VariantDef::default().name,
+                        toggle.default_variant.swap(0, Ordering::Relaxed),
+                    );
+
+                    (toggle.name.clone(), ToggleStats { yes, no, variants })
+                })
+                .collect();
+            let timestamp = Utc::now();
+            let bucket = MetricBucket {
+                toggles: metrics,
+                start: self.started,
+                stop: timestamp,
+            };
+            self.started = timestamp;
+            Some(bucket)
+        } else {
+            None
+        }
+    }
+
+    pub fn get_metrics(&mut self) -> Option<MetricBucket> {
+        self.harvest_metrics()
+    }
+
+    fn enabled(&self, toggle: &CompiledToggle, context: &Context) -> bool {
+        let context = EnrichedContext::from(context.clone(), toggle.name.clone());
+        let is_enabled = toggle.enabled && (toggle.compiled_strategy)(&context);
+        if is_enabled {
+            toggle.yes.fetch_add(1, Ordering::Relaxed);
         } else {
-            false
+            toggle.no.fetch_add(1, Ordering::Relaxed);
         }
+        is_enabled
     }
 
     pub fn is_enabled(&self, name: String, context: &Context) -> bool {
-        match &self.compiled_state {
-            Some(_) => {
-                let toggle = self.get_toggle(name);
-                self.enabled(toggle, context)
+        self.compiled_state
+            .as_ref()
+            .and_then(|_| {
+                self.get_toggle(name)
+                    .map(|toggle| self.enabled(toggle, context))
+            })
+            .unwrap_or_default()
+    }
+
+    fn resolve_variant<'a>(
+        &self,
+        toggle: &CompiledToggle,
+        variants: &'a Vec<CompiledVariant>,
+        context: &Context,
+    ) -> Option<&'a CompiledVariant> {
+        if variants.is_empty() {
+            return None;
+        }
+        if let Some(found_override) = check_for_variant_override(variants, context) {
+            return Some(found_override);
+        }
+        let total_weight: u32 = variants.iter().map(|var| var.weight as u32).sum();
```

Comment:
Seems like we're `as u32`ing the weights here. Related to https://github.com/Unleash/yggdrasil/pull/25/files#r1104638297?